python/ray/spark/__init__.py

```diff
@@ -0,0 +1,593 @@
+            ]
+
+            available_physical_gpus = get_spark_task_assigned_physical_gpus(
+                gpu_addr_list
+            )
+            ray_worker_cmd.append(
+                f"--num-gpus={len(available_physical_gpus)}",
+            )
+            ray_worker_extra_envs["CUDA_VISIBLE_DEVICES"] = ",".join(
+                [str(gpu_id) for gpu_id in available_physical_gpus]
+            )
+
+        if sys.platform.startswith("linux"):
+
+            def setup_sigterm_on_parent_death():
+                """
+                Uses prctl to automatically send SIGTERM to the command process when its parent is
+                dead.
+
+                This handles the case when the parent is a PySpark worker process.
+                If a user cancels the PySpark job, the worker process gets killed, regardless of
+                PySpark daemon and worker reuse settings.
+                We use prctl to ensure the command process receives SIGTERM after spark job
+                cancellation.
+                The command process itself should handle SIGTERM properly.
+                This is a no-op on macOS because prctl is not supported.
+
+                Note:
+                When a pyspark job cancelled, the UDF python process are killed by signal "SIGKILL",
+                This case neither "atexit" nor signal handler can capture SIGKILL signal.
+                prctl is the only way to capture SIGKILL signal.
+                """
+                try:
+                    import ctypes
+                    import signal
+
+                    libc = ctypes.CDLL("libc.so.6")
+                    # Set the parent process death signal of the command process to SIGTERM.
+                    libc.prctl(1, signal.SIGTERM)  # PR_SET_PDEATHSIG, see prctl.h
+                except OSError as e:
+                    _worker_logger.warning(
+                        f"Setup libc.prctl PR_SET_PDEATHSIG failed, error {repr(e)}."
+                    )
+
+        else:
+            setup_sigterm_on_parent_death = None
+
+        _worker_logger.info(f"Start Ray worker, command: {' '.join(ray_worker_cmd)}")
+
+        # TODO: write to NFS mount
+        with open(
+            os.path.join(ray_log_dir, f"ray-start-worker-{task_id}.log"),
+            "w",
+            buffering=1,
+        ) as worker_log_fp:
+            exec_cmd(
+                ray_worker_cmd,
+                synchronous=True,
+                capture_output=False,
+                stream_output=False,
+                extra_env=ray_worker_extra_envs,
+                preexec_fn=setup_sigterm_on_parent_death,
+                stdout=worker_log_fp,
+                stderr=subprocess.STDOUT,
+            )
+
+        # Delete the worker temp and log directories at the conclusion of running the
+        # submitted task.
+        shutil.rmtree(ray_temp_dir, ignore_errors=True)
+        shutil.rmtree(ray_log_dir, ignore_errors=True)
+
+        # NB: Not reachable.
```

Comment:
btw, what's NB?

Reply:
Latin shorthand for 'nota bene' ("note well", aka: this is potentially more important than other notes in the code)